backend/src/layer/logic/mealplan_management/relation_resolver.rs

```diff
@@ -0,0 +1,254 @@
+        };
+
+        for line in canteen.lines {
+            let name = &line.name.clone();
+            if let Err(_e) = self.handle_line(line, date).await {
+                warn!("Skip line '{:?}' as it could not be resolved", name);
+            }
+        }
+        Ok(())
+    }
+
+    async fn handle_line(&self, line: ParseLine, date: Date) -> Result<(), DataError> {
+        let db_line = match self.db.get_similar_line(&line.name).await? {
+            Some(similar_line) => self.db.update_line(similar_line.id, &line.name).await?,
+            None => self.db.insert_line(&line.name).await?,
+        };
+
+        let average = Self::determine_average_price(line.dishes.iter(), line.dishes.len())?;
+
+        for dish in line.dishes {
+            let name = &dish.name.clone();
+            if let Err(_e) = self.handle_dish(&db_line, dish, date, average).await {
+                warn!("Skip dish '{:?}' as it could not be resolved", name);
+            }
+        }
+        Ok(())
+    }
+
+    async fn handle_dish(
+        &self,
+        db_line: &Line,
+        dish: Dish,
+        date: Date,
+        average: u32,
+    ) -> Result<(), DataError> {
+        let similar_meal_result = self.db.get_similar_meal(&dish.name).await?;
+        let similar_side_result = self.db.get_similar_side(&dish.name).await?;
+        let price_limit = f64::from(average) * Self::PERCENTAGE;
+
+        // Case: A similar side and meal could be found. Uncommon case.
+        // Case: Just a meal could be found.
+        if let Some(similar_meal) = similar_meal_result {
+            self.db
+                .update_meal(similar_meal.id, db_line.id, date, &dish.name, &dish.price)
+                .await?;
+        // Case: A similar side could be found
+        } else if let Some(similar_side) = similar_side_result {
+            self.db
+                .update_side(similar_side.id, db_line.id, date, &dish.name, &dish.price)
+                .await?;
+        // Case: No similar meal could be found. Dish needs to be determined
+        } else if (f64::from(dish.price.price_student)) < price_limit
+            && !dish.name.contains(Self::get_edge_case_meal())
+        {
+            self.db
+                .insert_side(
+                    &dish.name,
+                    dish.meal_type,
+                    &dish.price,
+                    date,
+                    &dish.allergens,
+                    &dish.additives,
+                )
+                .await?;
+        } else {
+            self.db
+                .insert_meal(
+                    &dish.name,
+                    dish.meal_type,
+                    &dish.price,
+                    date,
+                    &dish.allergens,
+                    &dish.additives,
+                )
+                .await?;
+        };
+        Ok(())
+    }
+
+    fn determine_average_price(dishes: Iter<Dish>, len: usize) -> Result<u32, DataError> {
```

Comment:
this can be done much nicer with a `iterator.sum()` and interatd `iterator.count()` and could probably be inlined. Also you may just want to force-convert the usize tu u32?

Reply:
maybe directly return an f64 as that is what you use later anyways?